File: AppSets-Android/share/src/main/java/xcj/app/share/http/service/AppSetsShareServiceImpl.kt
package xcj.app.share.http.service

import android.content.Context
import xcj.app.share.base.ClientInfo
import xcj.app.share.base.ClientSendDataInfo
import xcj.app.share.base.DataContent
import xcj.app.share.base.DataProgressInfoPool
import xcj.app.share.base.ShareSystem
import xcj.app.share.http.HttpShareMethod
import xcj.app.share.ui.compose.AppSetsShareActivity
import xcj.app.starter.android.util.PurpleLogger
import xcj.app.starter.foundation.http.DesignResponse
import xcj.app.starter.util.ContentType
import xcj.app.web.webserver.base.FileUploadN
import java.util.UUID

class AppSetsShareServiceImpl : AppSetsShareService {

    companion object {
        private const val TAG = "AppSetsShareServiceImpl"
    }

    override fun greeting(context: Context, clientHost: String): DesignResponse<String> {
        return DesignResponse(data = "Hello, AppSets Share!")
    }

    override fun ping(context: Context, clientHost: String): DesignResponse<String> {
        return DesignResponse(data = "pong")
    }

    override fun isNeedPin(context: Context, clientHost: String): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        val isNeedPin = shareMethod.isNeedPin(clientInfo)
        PurpleLogger.current.d(
            TAG,
            "isNeedPin, clientHost:$clientHost, isNeedPin:$isNeedPin"
        )
        return DesignResponse(data = isNeedPin)
    }

    override fun pair(
        context: Context,
        clientHost: String,
        pin: Int
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        PurpleLogger.current.d(TAG, "pair, clientHost:$clientHost, pin:$pin")
        shareMethod.onClientRequestPair(pin, clientInfo)
        return DesignResponse(data = true)
    }

    override fun pairResponse(
        context: Context,
        clientHost: String,
        token: String
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        PurpleLogger.current.d(
            TAG,
            "pairResponse, clientHost:$clientHost, token:$token"
        )
        val clientInfo = ClientInfo(clientHost)
        shareMethod.onSeverPairResponse(token, clientInfo)
        return DesignResponse(data = true)
    }

    override fun prepareSend(
        context: Context,
        clientHost: String,
        token: String
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        val clientSendDataInfo = ClientSendDataInfo(emptyList())
        PurpleLogger.current.d(
            TAG,
            "prepareSend, clientHost:$clientHost, token:$token"
        )
        shareMethod.onClientPrepareSend(clientInfo, clientSendDataInfo)
        return DesignResponse(data = true)
    }

    override fun prepareSendResponse(
        context: Context,
        clientHost: String,
        token: String,
        isAccept: Boolean
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        PurpleLogger.current.d(
            TAG,
            "prepareSendResponse, clientHost:$clientHost, token:$token"
        )
        shareMethod.onServerPrepareSendResponse(clientInfo, isAccept)
        return DesignResponse(data = true)
    }

    override fun postText(
        context: Context,
        clientHost: String,
        token: String,
        text: String
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        val uuid = UUID.randomUUID().toString()
        val dataProgressInfo = DataProgressInfoPool.obtainById(uuid)
        dataProgressInfo.name = null
        val textLength = text.length.toLong()
        dataProgressInfo.total = textLength
        dataProgressInfo.current = textLength
        dataProgressInfo.percentage = 100.0
        PurpleLogger.current.d(
            TAG,
            "postText, clientHost:$clientHost, token:$token, text:$text"
        )
        shareMethod.dataReceivedProgressListener.onProgress(dataProgressInfo)
        shareMethod.onContentReceived(
            ContentType.APPLICATION_TEXT,
            DataContent.StringContent(text, clientInfo)
        )
        return DesignResponse(data = true)
    }

    override fun postFile(
        context: Context,
        clientHost: String,
        token: String,
        fileUploadN: FileUploadN
    ): DesignResponse<Boolean> {
        if (context !is AppSetsShareActivity) {
            return DesignResponse(data = false)
        }
        val shareMethod = context.getShareMethod()
        if (shareMethod !is HttpShareMethod) {
            return DesignResponse(data = false)
        }
        val clientInfo = ClientInfo(clientHost)
        PurpleLogger.current.d(
            TAG,
            "postFile, clientHost:$clientHost, token:$token, fileUploadN:${fileUploadN}"
        )
        saveFileIfNeeded(shareMethod, clientInfo, fileUploadN)

        return DesignResponse(data = true)
    }

    fun saveFileIfNeeded(
        shareMethod: HttpShareMethod,
        clientInfo: ClientInfo,
        fileUploadN: FileUploadN?
    ) {
        PurpleLogger.current.d(TAG, "saveFileIfNeeded, fileCount:${fileUploadN?.amount}")
        var currentFileUploadN: FileUploadN? = fileUploadN
        do {
            if (currentFileUploadN == null) {
                return
            }
            val fileUpload = currentFileUploadN.fileUpload

            if (fileUpload.isCompleted) {
                val fileToSave =
                    ShareSystem.makeFileIfNeeded(fileUpload.filename, createFile = false)
                if (fileToSave == null) {
                    continue
                }
                fileUpload.renameTo(fileToSave)

                val fileId = UUID.randomUUID().toString()
                val dataProgressInfoForEnd = DataProgressInfoPool.makeEnd(fileId)
                shareMethod.dataReceivedProgressListener.onProgress(dataProgressInfoForEnd)

                shareMethod.onContentReceived(
                    ContentType.APPLICATION_FILE,
                    DataContent.FileContent(fileUpload.file, null, clientInfo)
                )
            }
            currentFileUploadN = currentFileUploadN.next
        } while (currentFileUploadN != null)
    }

}

File: AppSets-iOS/AppSets/com/app/appsets/share/http/service/AppSetsShareServiceImpl.swift
//
//  AppSetsShareService.swift
//  AppSets
//
//  Created by caiju Xu on 2025/3/3.
//

struct AppSetsShareServiceImpl: AppSetsShareService {
    private static let TAG = "AppSetsShareServiceImpl"

    func greeting(context: any Context, clientHost: String) async -> DesignResponse<String> {
        PurpleLogger.current.d(AppSetsShareServiceImpl.TAG, "greeting")
        return DesignResponse(code: 0, data: "Hello, AppSets Share!")
    }

    func ping(context: any Context, clientHost: String) async -> DesignResponse<String> {
        PurpleLogger.current.d(AppSetsShareServiceImpl.TAG, "ping")
        return DesignResponse(code: 0, data: "pong")
    }

    func isNeedPin(context: any Context, clientHost: String) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func pair(context: any Context, clientHost: String, pin: Int) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func pairResponse(context: any Context, clientHost: String, shareToken: String) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func prepareSend(context: any Context, clientHost: String, shareToken: String, uri: String) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func prepareSendResponse(context: any Context, clientHost: String, shareToken: String, isAccept: Bool, preferDownloadSelf: Bool) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func postText(context: any Context, clientHost: String, shareToken: String, text: String) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func postFile(context: any Context, clientHost: String, shareToken: String, fileUploadN: FileUploadN) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func postFileChunked(context: any Context, clientHost: String, shareToken: String, fileId: String, chunkCount: Int, chunk: Int, fileUploadN: FileUploadN) async -> DesignResponse<Bool> {
        return DesignResponse(code: 0, data: true)
    }

    func getContent(context: any Context, clientHost: String, shareToken: String, contentId: String) async -> DesignResponse<ContentDownloadN> {
        let contentDownloadN = ContentDownloadN(id: "", name: "")
        return DesignResponse(code: 0, data: contentDownloadN)
    }

    func getContentList(context: any Context, clientHost: String, shareToken: String, uri: String) async -> DesignResponse<ContentInfoList> {
        PurpleLogger.current.d(AppSetsShareServiceImpl.TAG, "getContentList")
        guard let dataContentList = findContentListForContentUri(uri) else {
            return DesignResponse(code: 0, data: nil)
        }
        let uri = uri
        var infoList: [ContentInfo] = []
        dataContentList.forEach{ dataContent in
            if(dataContent is StringDataContent){
                let stringDataContent = dataContent as! StringDataContent
                let name = stringDataContent.name.data(using: .utf8)?.base64EncodedString() ?? ""
                let contentInfo = ContentInfo(id:stringDataContent.id, name: name, size: stringDataContent.content.count, type: ContentInfo.TYPE_STRING)
                infoList.append(contentInfo)
            }else if(dataContent is UriDataContent){
                let uriDataContent = dataContent as! UriDataContent
                let name = uriDataContent.name.data(using: .utf8)?.base64EncodedString() ?? ""
                let contentInfo = ContentInfo(id:uriDataContent.id, name: name, size: 0, type: ContentInfo.TYPE_URI)
                infoList.append(contentInfo)
            }
        }

        let contentInfoListWrapper = ContentInfoList(uri: uri, count: infoList.count, infoList: infoList)
        return DesignResponse(code: 0, data: contentInfoListWrapper)
    }

    func exchangeDeviceInfo(context: any Context, clientHost: String, device: HttpShareDevice) async -> DesignResponse<HttpShareDevice> {
        let deviceName = DeviceName.NONE
        let deviceAddress = DevcieAddress.NONE
        let device = HttpShareDevice(
             deviceName: deviceName,
             deviceAddress: deviceAddress,
             deviceType: ShareDevice.DEVICE_TYPE_PHONE
        )
        return DesignResponse(code: 0, data: device)
    }
    
    private func findContentListForContentUri(_ uri:String)->[any DataContent]? {
        return ShareViewModel.INSTANCE.getPendingSendContentList()
    }
}
